teps/0141-platform-context-variables.md

```diff
@@ -221,190 +232,319 @@ The downside of this approach is that Tasks can't easily reference the values of
 
 ## Proposal
 
-<!--
-This is where we get down to the specifics of what the proposal actually is.
-This should have enough detail that reviewers can understand exactly what
-you're proposing, but should not include things like API designs or
-implementation. The "Design Details" section below is for the real
-nitty-gritty.
--->
+There are several components to this proposal:
 
-### Notes and Caveats
+- New TaskRun and PipelineRun syntax allowing TaskRun and PipelineRun authors
+to reference platform-provided context variables
+- The mechanism platform builders use to specify the values of each context variable per Run
 
-<!--
-(optional)
+## Design Details
 
-Go in to as much detail as necessary here.
-- What are the caveats to the proposal?
-- What are some important details that didn't come across above?
-- What are the core concepts and how do they relate?
--->
+### Stability level
 
+This feature will start as an alpha feature and will be enabled by a new feature flag "enable-platform-context-variables".
 
-## Design Details
+### TaskRun and PipelineRun syntax
 
-<!--
-This section should contain enough information that the specifics of your
-change are understandable. This may include API specs (though not always
-required) or even code snippets. If there's any ambiguity about HOW your
-proposal will be implemented, this is the place to discuss them.
-
-If it's helpful to include workflow diagrams or any other related images,
-add them under "/teps/images/". It's upto the TEP author to choose the name
-of the file, but general guidance is to include at least TEP number in the
-file name, for example, "/teps/images/NNNN-workflow.jpg".
--->
+This TEP proposes adding support for the context variable substitutions in TaskRuns and PipelineRuns of the form `$(context.platform.foo)`, where `foo` is the name of a parameter supplied by the platform. Platforms may provide string, array, or object parameters, so TaskRuns and PipelineRuns may also use substitutions like `$(context.platform.my-array-param[*])` or `$(context.platform.my-object-param.key)`.
+
+For example, for the clone + build Pipeline referenced above, a user could define the following PipelineRun:
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: PipelineRun
+metadata:
+  name: clone-kaniko-build-push-run
+spec:
+  context:
```

Comment:
is the proposal about introducing a new section called `context` in `pipelineRun.spec` to define platform context? How is this different from a `param` section as can a param `repo-url` reference a context variable without defining it in a `context` section?

Are there any pre-defined context variables available for the users to run with? for example, as a platform provider, I have defined (somewhere out side of a `pipelineRun`) a few options for the users to choose from. The available options are:

```
STORAGECLASS-A: class-A,
STORAGECLASS-B: class-B,
STORAGECLASS-C: class-C
```

Will the pipeline author be able to utilize `STORAGECLASS-A` variable and tekton controller will be able to replace this with the right value - class-A?

Reply:
> is the proposal about introducing a new section called context in pipelineRun.spec to define platform context?

Yes. The field is used to accept the platform context. We add this field to PipelineRun.Spec, and platform can populate the values when creating the PipelineRun

> Are there any pre-defined context variables available for the users to run with?

Yes. And I think this is up to the platforms if they want to have pre-defined values. They need to add code to populate those values to pipelinerun, and users can use $(context.platform.STORAGECLASS-A) to use class-A.   

$(context.platform.STORAGECLASS-A) the syntax can also be customized by the platform. I added an optional section for this feature

Reply:
I see, how about putting it our of spec, and making it like this:
```
apiVersion: tekton.dev/v1
kind: PipelineRun
metadata:
  name: clone-kaniko-build-push-run
context:
  params:
  - name: repo_url
    value: https://github.com/tektoncd/pipeline
  - name: image_registry
    value: gcr.io/user
spec:
  pipelineRef:
    name: clone-kaniko-build-push
  params:
  - name: repo-url
    value: $(context.platform.repo_url)
  - name: image
    value: $(context.platform.image_registry)/myapp
```

> what's the benefit over just using params only 

The benefits are: 1) platform params won't have conflicts with users' params 2) provenance generator(Chains) would be able to tell which are internal params (platform params), which are external params (users params)